### PR TITLE
[ENH] :sparkles: LIST_GROUP of GenericLdapMapper should take ASI_GROUPS

### PR DESCRIPTION
**Pull request by mehdimaaref7**
Signed-off-by: Mehdi Maaref <mehdimaaref@live.fr>

### CODE REVIEW

File: sugoi-api-ldap-utils/src/main/java/fr/insee/sugoi/ldap/utils/mapper/GenericLdapMapper.java [hunk 1]
```diff
@@ -197,15 +197,25 @@ private static Object transformLdapAttributeToSugoiAttribute(
...
+                              .replace("{appliname}", "(.*)"));
                   Matcher matcher =
                       pattern.matcher(attributeValue.substring(attributeValue.indexOf(",") + 1));
+                  Matcher matcher_ASI = pattern_ASI.matcher(attributeValue);
                   if (matcher.matches()) {
                     return new Group(
                         matcher.group(1), LdapUtils.getNodeValueFromDN(attributeValue));
+                  } else if (matcher_ASI.matches()) {
+                    return new Group(
+                        matcher_ASI.group(1), LdapUtils.getNodeValueFromDN(attributeValue));
```
Comment (CChemin) [inline sugoi-api-ldap-utils/src/main/java/fr/insee/sugoi/ldap/utils/mapper/GenericLdapMapper.java:213]: remove all "_" in variable names

File: sugoi-api-ldap-utils/src/main/java/fr/insee/sugoi/ldap/utils/mapper/GenericLdapMapper.java [hunk 2]
```diff
@@ -197,15 +197,25 @@ private static Object transformLdapAttributeToSugoiAttribute(
...
                   } else {
                     return null;
                   }
                 })
+            .filter(a -> a != null)
```
Comment (CChemin) [inline sugoi-api-ldap-utils/src/main/java/fr/insee/sugoi/ldap/utils/mapper/GenericLdapMapper.java:218]: you can use Objects::nonNull

File: sugoi-api-ldap-utils/src/main/java/fr/insee/sugoi/ldap/utils/mapper/GenericLdapMapper.java [hunk 3]
```diff
@@ -197,15 +197,25 @@ private static Object transformLdapAttributeToSugoiAttribute(
                           config
                               .get(LdapConfigKeys.GROUP_SOURCE_PATTERN)
                               .replace("{appliname}", "(.*)"));
+                  Pattern pattern_ASI =
```
Comment (CChemin) [inline sugoi-api-ldap-utils/src/main/java/fr/insee/sugoi/ldap/utils/mapper/GenericLdapMapper.java:200]: move the pattern out of the map

**Comment by mehdimaaref7** [comment]
Done ✅ @CChemin :) you can merge this PR after these PRs : #731 #730

File: sugoi-api-ldap-utils/src/main/java/fr/insee/sugoi/ldap/utils/mapper/GenericLdapMapper.java [hunk 4]
```diff
@@ -189,6 +185,11 @@ private static Object transformLdapAttributeToSugoiAttribute(
       case LIST_GROUP:
         values = new ArrayList<>();
```
Comment (CChemin) [inline sugoi-api-ldap-utils/src/main/java/fr/insee/sugoi/ldap/utils/mapper/GenericLdapMapper.java:186]: you can simplify these lines by using a flatmap like : 
`        return attrs.stream().flatMap(attribute -> Arrays.stream(attribute.getValues()))
`

File: docs/configuration.md [hunk 5]
```diff
@@ -92,28 +92,28 @@ Sugoi-api implements springdoc with full customization allowed
...
+| fr.insee.sugoi.api.old.enable.preauthorize                 |                                                                                                                        |               |         |
+| fr.insee.sugoi.api.regexp.role.reader                      |                                                                                                                        |               |         |
+| fr.insee.sugoi.api.regexp.role.writer                      |                                                                                                                        |               |         |
+| fr.insee.sugoi.api.regexp.role.admin                       |                                                                                                                        |               |         |
+| fr.insee.sugoi.api.regexp.role.app.manager                 | Default pattern to use when searching manager group for application. Application name should be passed via {appliname} |               |         |
```
Comment (CChemin) [inline docs/configuration.md:113]: I mistook this variable with the variable `fr.insee.sugoi.ldap.default.group_manager_source_pattern` . Can you move the description to this variable and change the default ref in realm-configuration.md ?

File: sugoi-api-ldap-utils/src/main/java/fr/insee/sugoi/ldap/utils/mapper/GenericLdapMapper.java (hunk 4 above)
Comment (mehdimaaref7) [reply sugoi-api-ldap-utils/src/main/java/fr/insee/sugoi/ldap/utils/mapper/GenericLdapMapper.java]: Done ✅ :) @CChemin

File: docs/configuration.md (hunk 5 above)
Comment (mehdimaaref7) [reply docs/configuration.md]: Done ✔️ @CChemin  👍

**Comment by mehdimaaref7** [comment]
Everything looks good, I added an optimization commit where some unnecessary lines were deleted. 👍

File: sugoi-api-ldap-utils/src/test/java/fr/insee/sugoi/ldap/utils/mapper/UserLdapMapperFromAttributesTest.java [hunk 6]
```diff
@@ -171,31 +158,47 @@ public void getUserGroupFromAttributes() {
...
-    attributes.add(groupAttributes1);
-    attributes.add(groupAttributes2);
+            "memberOf", "cn=reader,ou=monappli2_Objets,ou=monappli2,ou=Applications,o=insee,c=fr");
+
+    Collection<Attribute> attributes = List.of(groupAttributes1, groupAttributes2);
     User mappedUser = userLdapMapper.mapFromAttributes(attributes);
 
     assertThat(
-        "Should have admin group",
+        "monappli App should have admin name",
```
Comment (clement-dufaure) [inline sugoi-api-ldap-utils/src/test/java/fr/insee/sugoi/ldap/utils/mapper/UserLdapMapperFromAttributesTest.java:167]: no, test checks if the user have admin group

File: sugoi-api-ldap-utils/src/test/java/fr/insee/sugoi/ldap/utils/mapper/UserLdapMapperFromAttributesTest.java [hunk 7]
```diff
@@ -171,31 +158,47 @@ public void getUserGroupFromAttributes() {
...
         mappedUser.getGroups().stream().anyMatch(group -> group.getName().equals("admin")));
     assertThat(
-        "Admin group should have monappli app name",
-        mappedUser.getGroups().stream().anyMatch(group -> group.getAppName().equals("monappli")));
+        "Reader group should have monappli2 app name",
```
Comment (clement-dufaure) [inline sugoi-api-ldap-utils/src/test/java/fr/insee/sugoi/ldap/utils/mapper/UserLdapMapperFromAttributesTest.java:170]: test checks if user has at least a group of monappli2 app

File: sugoi-api-ldap-utils/src/test/java/fr/insee/sugoi/ldap/utils/mapper/UserLdapMapperFromAttributesTest.java [hunk 8]
```diff
@@ -171,31 +158,47 @@ public void getUserGroupFromAttributes() {
...
+        new Attribute("memberOf", "cn=ASI_toto,ou=toto,ou=Applications,o=insee,c=fr");
+
+    Attribute groupAttributes =
+        new Attribute(
+            "memberOf", "cn=admin,ou=monappli_Objets,ou=monappli,ou=Applications,o=insee,c=fr");
+    Collection<Attribute> attributes = List.of(groupAttributesASI, groupAttributes);
+    User mappedUser = userLdapMapper.mapFromAttributes(attributes);
+
+    assertThat(
+        "Manager group should have group ASI_toto name",
```
Comment (clement-dufaure) [inline sugoi-api-ldap-utils/src/test/java/fr/insee/sugoi/ldap/utils/mapper/UserLdapMapperFromAttributesTest.java:186]: check test message

**Review by clement-dufaure — APPROVED** [review]

**Pull request merged**